### PR TITLE
add codecov json format (support region testing)

### LINKED ISSUE #20: Support region coverage on codecov
readme says:

> Currently, only line coverage is available on Codecov. This is because `-Z instrument-coverage` does not support branch coverage and Codecov does not support region coverage.

However, we can probably support region coverage on codecov by converting llvm-cov's json coverage format to [codecov's json coverage format](https://docs.codecov.io/docs/codecov-custom-coverage-format).

related:
- https://llvm.org/docs/CoverageMappingFormat.html

### PR DESCRIPTION
**Pull request by andrewgazelka**
Fixes #20

### CODE REVIEW

File: tests/auxiliary/mod.rs [hunk 1]
```diff
@@ -90,7 +91,13 @@ pub fn normalize_output(output_path: &Utf8Path, args: &[&str]) -> Result<()> {
...
             json.demangle();
         }
-        fs::write(output_path, serde_json::to_vec_pretty(&json)?)?;
+        let json = if args.contains(&"--codecov") {
+            let codecov = CodeCovJsonExport::from(json);
+            serde_json::to_vec_pretty(&codecov)?
+        } else {
+            serde_json::to_vec_pretty(&json)?
+        };
+        fs::write(output_path, json)?;
```
Comment (taiki-e) [inline tests/auxiliary/mod.rs:100]: This function is for sanitizing platform-specific paths and demangling function names in the reports generated during testing.

For argument handling and report generation, https://github.com/taiki-e/cargo-llvm-cov/pull/224, which supported Cobertura format in a similar way, may be helpful.
(Basically,  `cli.rs` [handles the arguments](https://github.com/taiki-e/cargo-llvm-cov/pull/224/files#diff-b2812f19576dd53d0c35b107a322f58a00fce6977f4b5976e1961853982af3cc), and `main.rs`'s `Format` enum [handle the format-specific code](https://github.com/taiki-e/cargo-llvm-cov/pull/224/files#diff-42cb6807ad74b3e201c5a7ca98b911c5fa08380e942be6e4ac5807f8377f87fc).)

**Review by taiki-e — CHANGES_REQUESTED** [review]
Thanks!

**Comment by andrewgazelka** [comment]
Does this look good?

https://github.com/andrewgazelka/cargo-llvm-cov/blob/ed1cb3a0263ba8759039a2bab0fb6cb8c45e8538/tests/fixtures/coverage-reports/real1/all.codecov.json#L1-L48

File: tests/auxiliary/mod.rs (hunk 1 above)
Comment (andrewgazelka) [reply tests/auxiliary/mod.rs]: removed this :)

**Comment by andrewgazelka** [comment]
Looks like it works (at least somewhat?) https://app.codecov.io/gh/getcollective-ai/collective/tree/region-ci

**Comment by andrewgazelka** [comment]
also it works _much_ better when testing in release mode

**Comment by andrewgazelka** [comment]
@taiki-e I rebased :)

File: src/main.rs [hunk 2]
```diff
@@ -868,7 +871,27 @@ impl Format {
...
+            let cov: LlvmCovJsonExport = serde_json::from_str(&cov)?;
+            let cov = CodeCovJsonExport::from(cov);
+            let out = serde_json::to_string(&cov)?;
+
+            if let Some(output_path) = &cx.args.cov.output_path {
+                fs::write(output_path, out)?;
+                eprintln!();
+                status!("Finished", "report saved to {output_path}");
+            } else {
+                // write XML to stdout
```
Comment (taiki-e) [inline src/main.rs:890]:
```suggestion
                // write JSON to stdout
```

File: src/json.rs [hunk 3]
```diff
@@ -15,6 +18,101 @@ pub struct LlvmCovJsonExport {
     pub(crate) version: String,
 }
 
+/// <https://docs.codecov.com/docs/codecov-custom-coverage-format>
+///
+/// The fraction
```
Comment (taiki-e) [inline src/json.rs:23]: This seems like a sentence in the process of writing.

Comment (andrewgazelka) [reply src/json.rs]: 0175abe

Comment (andrewgazelka) [reply src/json.rs]: https://github.com/taiki-e/cargo-llvm-cov/blob/0175abeca07d49f820d0b2df50e927d3f0976372/src/json.rs#L21-L28

**Review by taiki-e — APPROVED** [review]
Looks great!

**Comment by taiki-e** [comment]
Published in 0.5.12.

**Pull request merged**